### PR TITLE
Refactor handling of single- and double-quoted strings

### LINKED ISSUE #11: Trouble with quoted string that starts with a `#` character
I think the parser is having trouble with strings starting with a `#` character. If you add the following test case, the parser will crash on it:

```c
test "single quoted string with a pound symbol" {
    const source =
        \\- hello
        \\- '#hello'
    ;

    var yaml = try Yaml.load(testing.allocator, source);
    defer yaml.deinit();

    const arr = try yaml.parse([2][]const u8);
    try testing.expectEqual(arr.len, 2);
    try testing.expect(mem.eql(u8, arr[0], "hello"));
    try testing.expect(mem.eql(u8, arr[1], "#hello"));
}
```
I *think* this is valid. I've been using zig-yaml to parse color schemes (with lots of RGB color hex codes), but after updating the project I ran into this issue.

### PR DESCRIPTION
**Pull request by kubkon**
Fixes #11 

**tl;dr** I figured the current approach for generically handling single- and double-quoted strings was not easily extensible, and instead of cramming tokenizing quoted phrases in the tokenizer, why not just defer parsing the contents of the quoted strings until the parser run? It seems like the approach implemented in this PR is simpler and cleaner and more easily extensible. It works like this:

1. tokenizer encodes the entire quoted string as a single token, either `.single_quoted` or `.double_quoted`, tracking any encountered escape sequence
2. parser parses the contents of the quoted string substituting requested characters, codepoints, etc. in place of the encountered escape sequences

cc @rudedogg |'ve re-used your test case from #11 (thank you!), and have added you as a co-author to the matching commit as a way of attributing that bit of code to you - hope that's OK!

### CODE REVIEW

File: src/Tokenizer.zig [hunk 1]
```diff
@@ -84,32 +76,36 @@ pub const TokenIterator = struct {
     }
 };
 
-fn matchesPattern(self: Tokenizer, comptime pattern: []const u8) bool {
+fn stringMatchesPattern(comptime pattern: []const u8, slice: []const u8) bool {
```
[the remark this answers is not in the document]
Comment (kubkon) [reply src/Tokenizer.zig:79]: Great question! Given that all patterns we want to search for are known during compilation time, we can unroll the loop into a sequence of `if`s which should generally be faster at runtime than looping. This is especially true in our case where patterns are generally really short, and this bit of code acts effectively as a short-hand for unrolling the loop manually into a series of `if`s. On the other hand, we can't call `std.mem.eql` at compilation time as we only know one input slice *a priori* (i.e., the *pattern*) while the other (the one we compare against) is only runtime known.
